### ISSUE TITLE
Failure in test_repair_disjoint_row_3nodes_diff_shard_count: got exception while processing a message: fmt::v7::format_error (invalid type specifier)

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#10880]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/41/testReport/repair_additional_test/TestRepairAdditional/Run_Dtest_Parallel_Cloud_Machines___HeavyDtest___heavy_split001___test_repair_disjoint_row_3nodes_diff_shard_count/

Scylla version `5.1.dev-0.20220625.78750c2e1a45`

```
self = <repair_additional_test.TestRepairAdditional object at 0x7f087b0a46d0>

    @pytest.mark.dtest_heavy
    def test_repair_disjoint_row_3nodes_diff_shard_count(self):
>       return self._repair_disjoint_row_3nodes_test(same_shard_count=False)

repair_additional_test.py:2844: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
repair_additional_test.py:2198: in _repair_disjoint_row_3nodes_test
    info = self._repair(node3, more_options + ['ks'])
repair_additional_test.py:228: in _repair
    return node.repair(options)
../scylla/.local/lib/python3.9/site-packages/ccmlib/node.py:1342: in repair
    return self.nodetool(cmd, **kwargs)
../scylla/.local/lib/python3.9/site-packages/ccmlib/scylla_node.py:684: in nodetool
    return super().nodetool(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ccmlib.scylla_node.ScyllaNode object at 0x7f087a2c4880>
cmd = 'repair ks', capture_output = True, wait = True, timeout = None

    def nodetool(self, cmd, capture_output=True, wait=True, timeout=None):
        """
        Setting wait=False makes it impossible to detect errors,
        if capture_output is also False. wait=False allows us to return
        while nodetool is still running.
        When wait=True, timeout may be set to a number, in seconds,
        to limit how long the function will wait for nodetool to complete.
        """
        if capture_output and not wait:
            raise common.ArgumentError("Cannot set capture_output while wait is False.")
        env = self.get_env()
        if self.is_scylla() and not self.is_docker():
            host = self.address()
        else:
            host = 'localhost'
        nodetool = self.get_tool('nodetool')
    
        if not isinstance(nodetool, list):
            nodetool = [nodetool]
        # see https://www.oracle.com/java/technologies/javase/8u331-relnotes.html#JDK-8278972
        nodetool.extend(['-h', host, '-p', str(self.jmx_port), '-Dcom.sun.jndi.rmiURLParsing=legacy'])
        nodetool.extend(cmd.split())
        if capture_output:
            p = subprocess.Popen(nodetool, universal_newlines=True, env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
            stdout, stderr = p.communicate(timeout=timeout)
        else:
            p = subprocess.Popen(nodetool, env=env, universal_newlines=True)
            stdout, stderr = None, None
    
        if wait:
            exit_status = p.wait(timeout=timeout)
            if exit_status != 0:
>               raise NodetoolError(" ".join(nodetool), exit_status, stdout, stderr)
E               ccmlib.node.NodetoolError: Nodetool command '/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.ccm/scylla-repository/78750c2e1a455d01e4f6b208760883b5b1794b5f/scylla-tools-java/bin/nodetool -h 127.0.11.3 -p 7199 -Dcom.sun.jndi.rmiURLParsing=legacy repair ks' failed; exit status: 2; stdout: [2022-06-25 04:02:18,227] Starting repair command #1, repairing 1 ranges for keyspace ks (parallelism=SEQUENTIAL, full=true)
E               [2022-06-25 06:30:34,766] Repair session 1 failed
E               [2022-06-25 06:30:34,767] Repair session 1 finished
E               ; stderr: error: Repair job has failed with the error message: [2022-06-25 06:30:34,766] Repair session 1 failed
E               -- StackTrace --
E               java.lang.RuntimeException: Repair job has failed with the error message: [2022-06-25 06:30:34,766] Repair session 1 failed
E               	at org.apache.cassandra.tools.RepairRunner.progress(RepairRunner.java:124)
E               	at org.apache.cassandra.utils.progress.jmx.JMXNotificationProgressListener.handleNotification(JMXNotificationProgressListener.java:77)
E               	at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.dispatchNotification(ClientNotifForwarder.java:583)
E               	at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.doRun(ClientNotifForwarder.java:533)
E               	at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.run(ClientNotifForwarder.java:452)
E               	at com.sun.jmx.remote.internal.ClientNotifForwarder$LinearExecutor$1.run(ClientNotifForwarder.java:108)

../scylla/.local/lib/python3.9/site-packages/ccmlib/node.py:795: NodetoolError
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/41/artifact/logs-heavy.release.001/dtest-gw3.log
```
04:02:17,268 396     repair_additional_test         DEBUG    repair_additional_test.py:2197 | test_repair_disjoint_row_3nodes_diff_shard_count: starting repair...
04:02:17,945 396     cassandra.cluster              INFO     thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Host 127.0.11.2:9042 may be up; will prepare queries and open connection pool
04:02:19,394 396     cassandra.cluster              INFO     thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Host 127.0.11.2:9042 may be up; will prepare queries and open connection pool
04:02:29,104 396     cassandra.pool                 INFO     thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Successful reconnection to 127.0.11.2:9042, marking node up if it isn't already
04:02:29,105 396     cassandra.cluster              INFO     thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Host 127.0.11.2:9042 may be up; will prepare queries and open connection pool
06:30:33,973 396     cassandra.cluster              WARNING  thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Host 127.0.11.1:9042 has been marked down
06:30:33,988 396     cassandra.cluster              WARNING  thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Host 127.0.11.1:9042 has been marked down
06:30:33,988 396     cassandra.cluster              WARNING  thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Host 127.0.11.1:9042 has been marked down
06:30:34,977 396     cassandra.pool                 WARNING  thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Error attempting to reconnect to 127.0.11.1:9042, scheduling retry in 2.18 seconds: [Errno 111] Tried connecting to [('127.0.11.1', 9042)]. Last error: Connection refused
06:30:34,992 396     cassandra.pool                 WARNING  thread.py           :58   | test_repair_disjoint_row_3nodes_diff_shard_count: Error attempting to reconnect to 127.0.11.1:9042, scheduling retry in 2.22 seconds: [Errno 111] Tried connecting to [('127.0.11.1', 9042)]. Last error: Connection refused
06:30:35,452 396     errors                         ERROR    conftest.py         :199  | test_repair_disjoint_row_3nodes_diff_shard_count: test failed: 
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/41/artifact/logs-heavy.release.001/1656138638665_repair_additional_test.py%3A%3ATestRepairAdditional%3A%3Atest_repair_disjoint_row_3nodes_diff_shard_count/node3.log
```
INFO  2022-06-25 04:02:18,220 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: starting user-requested repair for keyspace ks, repair id 1, options {{ trace -> false}, { jobThreads -> 1}, { incremental -> false}, { parallelism -> parallel}, { primaryRange -> false}}
INFO  2022-06-25 04:02:18,220 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: Skipped sending repair_flush_hints_batchlog to nodes=[127.0.11.1, 127.0.11.2, 127.0.11.3]
INFO  2022-06-25 04:02:18,220 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: Started to repair 1 out of 1 tables in keyspace=ks, table=cf, table_id=4fae3ef0-f43b-11ec-ac89-3c19cdbf7348, repair_reason=repair
INFO  2022-06-25 04:02:18,221 [shard 1] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: Started to repair 1 out of 1 tables in keyspace=ks, table=cf, table_id=4fae3ef0-f43b-11ec-ac89-3c19cdbf7348, repair_reason=repair
INFO  2022-06-25 04:02:18,221 [shard 2] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: Started to repair 1 out of 1 tables in keyspace=ks, table=cf, table_id=4fae3ef0-f43b-11ec-ac89-3c19cdbf7348, repair_reason=repair
INFO  2022-06-25 04:02:21,788 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: shard 0 stats: repair_reason=repair, keyspace=ks, tables={cf}, ranges_nr=769, round_nr=1474, round_nr_fast_path_already_synced=769, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=705, rpc_call_nr=12715, tx_hashes_nr=0, rx_hashes_nr=1531, duration=3.5680888 seconds, tx_row_nr=1354, rx_row_nr=658, tx_row_bytes=254552, rx_row_bytes=123704, row_from_disk_bytes={{127.0.11.1, 65048}, {127.0.11.2, 58656}, {127.0.11.3, 65424}}, row_from_disk_nr={{127.0.11.1, 346}, {127.0.11.2, 312}, {127.0.11.3, 348}}, row_from_disk_bytes_per_sec={{127.0.11.1, 0.0173859}, {127.0.11.2, 0.0156775}, {127.0.11.3, 0.0174864}} MiB/s, row_from_disk_rows_per_sec={{127.0.11.1, 96.9707}, {127.0.11.2, 87.4418}, {127.0.11.3, 97.5312}} Rows/s, tx_row_nr_peer={{127.0.11.1, 660}, {127.0.11.2, 694}}, rx_row_nr_peer={{127.0.11.1, 346}, {127.0.11.2, 312}}
INFO  2022-06-25 04:02:21,788 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: shard 0 completed successfully, keyspace=ks
INFO  2022-06-25 04:02:22,378 [shard 2] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: shard 2 stats: repair_reason=repair, keyspace=ks, tables={cf}, ranges_nr=769, round_nr=1483, round_nr_fast_path_already_synced=769, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=714, rpc_call_nr=12808, tx_hashes_nr=0, rx_hashes_nr=1538, duration=4.157466 seconds, tx_row_nr=1340, rx_row_nr=646, tx_row_bytes=251920, rx_row_bytes=121448, row_from_disk_bytes={{127.0.11.1, 58468}, {127.0.11.2, 62980}, {127.0.11.3, 65236}}, row_from_disk_nr={{127.0.11.1, 311}, {127.0.11.2, 335}, {127.0.11.3, 347}}, row_from_disk_bytes_per_sec={{127.0.11.1, 0.0134119}, {127.0.11.2, 0.0144469}, {127.0.11.3, 0.0149644}} MiB/s, row_from_disk_rows_per_sec={{127.0.11.1, 74.8052}, {127.0.11.2, 80.5779}, {127.0.11.3, 83.4643}} Rows/s, tx_row_nr_peer={{127.0.11.1, 682}, {127.0.11.2, 658}}, rx_row_nr_peer={{127.0.11.1, 311}, {127.0.11.2, 335}}
INFO  2022-06-25 04:02:22,378 [shard 2] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: shard 2 completed successfully, keyspace=ks
INFO  2022-06-25 06:30:33,987 [shard 0] gossip - Got shutdown message from 127.0.11.1, received_generation=1656129687, local_generation=1656129687
INFO  2022-06-25 06:30:33,987 [shard 0] gossip - InetAddress 127.0.11.1 is now DOWN, status = shutdown
INFO  2022-06-25 06:30:33,990 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: Started to shutdown off-strategy compaction updater
INFO  2022-06-25 06:30:33,990 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: Finished to shutdown off-strategy compaction updater
WARN  2022-06-25 06:30:33,990 [shard 0] repair - repair[654cc875-e216-4713-b3ca-ecc4688d7ef9]: repair_tracker run failed: std::runtime_error ({shard 1: std::_Nested_exception<std::__exception_ptr::exception_ptr>: seastar::rpc::closed_error (connection is closed)})
INFO  2022-06-25 06:30:35,624 [shard 0] compaction_manager - Asked to stop
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/41/artifact/logs-heavy.release.001/1656138638665_repair_additional_test.py%3A%3ATestRepairAdditional%3A%3Atest_repair_disjoint_row_3nodes_diff_shard_count/node1.log
```
INFO  2022-06-25 04:01:53,578 [shard 0] gossip - InetAddress 127.0.11.2 is now UP, status = NORMAL
INFO  2022-06-25 04:01:53,579 [shard 0] storage_service - Node 127.0.11.2 state jump to normal
INFO  2022-06-25 04:01:53,593 [shard 0] compaction - [Compact system.peers 8c4fca90-f43b-11ec-bdd0-0a87f9188128] Compacting [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-tyd0099e/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-28-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-tyd0099e/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-26-big-Data.db:level=0:origin=compaction]
INFO  2022-06-25 04:01:53,598 [shard 0] compaction - [Compact system.peers 8c4fca90-f43b-11ec-bdd0-0a87f9188128] Compacted 2 sstables to [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-tyd0099e/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-30-big-Data.db:level=0]. 90kB to 45kB (~50% of original) in 4ms = 11MB/s. ~256 total partitions merged to 1.
INFO  2022-06-25 04:02:05,861 [shard 0] storage_service - Node 127.0.11.2 state jump to normal
INFO  2022-06-25 04:02:05,875 [shard 0] compaction - [Compact system.peers 93a1b920-f43b-11ec-bdd0-0a87f9188128] Compacting [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-tyd0099e/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-32-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-tyd0099e/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-30-big-Data.db:level=0:origin=compaction]
INFO  2022-06-25 04:02:05,878 [shard 0] compaction - [Compact system.peers 93a1b920-f43b-11ec-bdd0-0a87f9188128] Compacted 2 sstables to [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-tyd0099e/test/node1/data/system/peers-37f71aca7dc2383ba70672528af04d4f/me-34-big-Data.db:level=0]. 90kB to 45kB (~50% of original) in 2ms = 22MB/s. ~256 total partitions merged to 1.
INFO  2022-06-25 04:02:22,594 [shard 0] rpc - client 127.0.11.3:52126 msg_id 6860:  got exception while processing a message: fmt::v7::format_error (invalid type specifier)
INFO  2022-06-25 06:30:33,972 [shard 0] compaction_manager - Asked to stop
```

**Comment by bhalevy** [comment]
@xemul could this be a result of c437920c23aac97d30fdba1e1c6554f15d98eb5a?
(although the format change in `send_unknown_verb_reply` looks benign)

**Comment by xemul** [comment]
@bhalevy , what kind of exception? There should also be https://github.com/scylladb/seastar/commit/ff46af9ae0714cdd1d42a1429e32de3eb01c239a applied not to let sending exception creep into continuation

**Comment by bhalevy** [comment]
@xemul it looks like the exception comes from a `fmt::format` call, that complains that the type specifier given to it in the format string is invalid.

**Comment by xemul** [comment]
Ah, I see. Yes, I dequeued the faulty patch Friday evening.

**Comment by bhalevy** [comment]
wait, but this test failed in master, not in next.
Maybe it's related to `[scylladb-dev] [PATCH v2] exceptions: Define operator<< for exception_code`?

**Comment by xemul** [comment]
> wait, but this test failed in master, not in next.

Ouch :(

> Maybe it's related to [scylladb-dev] [PATCH v2] exceptions: Define operator<< for exception_code?

This is a compilation patch, without it scylla won't compile in some cases. Well _maybe_ the current formatter doesn't format enums with `:d`, but as far as I can tell, this code is triggered by unit tests, so we'd face this early

**Comment by bhalevy** [comment]
Another data point: https://jenkins.scylladb.com/view/master/job/scylla-master/job/next/5043/testReport/junit/cql_tests/TestStorageProxyCQL/test_statements/
```
cassandra.cluster.NoHostAvailable: ('Unable to connect to any servers', {'127.0.34.1:9042': ConnectionException('Did not get expected SupportedMessage response; instead, got: <Error from server: code=0000 [Server error] message="invalid type specifier">')})
```

**Comment by xemul** [comment]
Yes, this comes from fmt

**Comment by slivne** [comment]
Lets revert the patch that exposed this - and apply it after we have a fix 

@xemul - what do we need to revert ? (/cc @avikivity )

@gleb-cloudius - according to the rumor this is in rpc level and we need your help here

**Comment by xemul** [comment]
> @xemul - what do we need to revert ? (/cc @avikivity )

:man_shrugging: , it was a suggestion that RPC send patchset could cause this. But the potential exception that could creep through connection::send() would render 'exceptional future ignore' warning, but not some other continuation get unexpected one (@slivne )

**Comment by xemul** [comment]
Found a suspicious place in `src/rpc/rpc.cc`

```c++
  xshard_connection_ptr connection::get_stream(connection_id id) const {
      auto it = _streams.find(id);
      if (it == _streams.end()) {
          throw std::logic_error(format("rpc stream id {:d} not found", id).c_str());
      }
      return it->second;
  }
```

The `connection_id` is a struct. It has its overload of `operator<<`, but formatting it with `:d` will throw, for example see f3841c1b457a8fffd14ebdb9e78beb4402e3c315, the same formatting for custom type called for `s/{:d}/{}/`

**Comment by xemul** [comment]
@bhalevy , I believe this can be closed as dup of #10950

**Comment by bhalevy** [comment]
This issue is still about the fmt exception on the server side, so it can be just closed with 98aa3ec99b9b16dc60bace1aa8982e8a7d58514f / https://github.com/scylladb/seastar/commit/15ec2d8bfdd8b45ba45ac69f541224005c89d276
